helix-core/src/syntax.rs

```diff
@@ -1946,6 +2098,116 @@ impl<'a> Iterator for HighlightIter<'a> {
     }
 }
 
+impl<'a> Iterator for RainbowIter<'a> {
+    type Item = Result<HighlightEvent, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        'main: loop {
+            // If we've already determined the next highlight boundary, just return it.
+            if let Some(e) = self.next_event.take() {
+                return Some(Ok(e));
+            }
+
+            // If the iterator has been cancelled, return an Error.
+            if let Err(err) = self.check_cancellation() {
+                return Some(Err(err));
+            }
+
+            // If none of the layers have any more highlight boundaries, terminate.
+            if self.layers.is_empty() {
+                return None;
+            }
+
+            if let Some(event) = self.scan_to_earliest_capture() {
+                return event;
+            };
+
+            let layer = &mut self.layers[0];
+            let (match_, capture_index) = layer.captures.next().unwrap();
+            let capture = match_.captures[capture_index];
+            let range = capture.node.byte_range();
+
+            // Remove from the rainbow scope stack any rainbow scopes that have already ended.
+            while let Some(scope) = self.context.rainbow_stack.last() {
+                if range.start >= scope.range.end {
+                    self.context.rainbow_stack.pop();
+                } else {
+                    break;
+                }
+            }
+
+            // If the node represents a rainbow scope, push a new rainbow scope onto
+            // the scope stack.
+            if Some(capture.index) == layer.config.rainbow_scope_capture_index {
+                let mut scope = RainbowScope {
+                    range: range.clone(),
+                    node_id: capture.node.id(),
+                    highlight: Highlight(
+                        self.context.rainbow_stack.len() % self.context.rainbow_length,
+                    ),
+                };
+                for prop in layer
+                    .config
+                    .rainbow_query
+                    .property_settings(match_.pattern_index)
+                {
+                    if prop.key.as_ref() == "rainbow.include-children" {
+                        scope.node_id = usize::MAX;
+                    }
+                }
+                self.context.rainbow_stack.push(scope);
+            }
```

Comment:
note: this is also mostly duplicate and could be in a function that calls back to a trait for a couple details